### PR TITLE
Module 8

### PR DESCRIPTION
**Pull request by SamSeed-Softwire**

### CODE REVIEW

**Comment by SamSeed-Softwire** [comment]
When Travis logs into Docker the following warning is displayed:
```
WARNING! Your password will be stored unencrypted in /home/travis/.docker/config.json.
Configure a credential helper to remove this warning. See
https://docs.docker.com/engine/reference/commandline/login/#credentials-store
```
I struggled to get this to work (I tried using the `pass` credential helped but no success - can't remember exactly where I got stuck).

How important is this to resolve?

File: Dockerfile [hunk 1]
```diff
@@ -58,8 +55,10 @@ CMD [ "--host=0.0.0.0" ]
 # Create an image used for running the app in a production environment.
 FROM base-with-app-code as prod
 
+RUN export FLASK_ENV=production
+
 # Define commands to be run when container is started.
-CMD [ "poetry", "run", "gunicorn", "--bind=0.0.0.0:5000", "--chdir", "./application", "app:create_app()" ]
+CMD poetry run gunicorn --bind=0.0.0.0:$PORT --chdir ./application 'app:create_app()'
```
Comment (JackMeadDev) [inline Dockerfile:65]: Note that this means you are insisting that a PORT environment variable is set - this means your local docker instructions for prod will only work if a PORT variable is set (and so following the README doesn't necessarily work any more)

We could fix that by setting a default environment variable in the Dockerfile (e.g. `ENV PORT=5000`) or by asking the user to pass one through in their `.env`

File: .travis.yml [hunk 2]
```diff
@@ -1,37 +1,83 @@
-#if: type = pull_request
+if: type = pull_request
```
Comment (JackMeadDev) [inline .travis.yml:1]: Worth noting that we don't want this always restricted to PRs, as we'll want deployments to run on the master branch

File: .travis.yml [hunk 3]
```diff
@@ -1,37 +1,83 @@
...
+    - docker build --target prod --cache-from samseedsoftwire/todo-app-prod:latest .
+    after_success:
+    - docker tag samseedsoftwire/todo-app-prod:latest registry.heroku.com/$HEROKU_APP_NAME/web
+    before_deploy:
+    - heroku container:login --verbose
+    deploy:
+      provider: script
+      script: bash ./deploy_prod.sh
+      on:
+        branch: module_8 # TODO: update this once I've sorted out master branch
```
Comment (JackMeadDev) [inline .travis.yml:44]: Don't forget this line on merge

File: travis_performance.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
+# Totals
```
Comment (JackMeadDev) [inline travis_performance.md:1]: This looks interesting! Was this manually compiled for analysis?

File: .travis.yml [hunk 5]
```diff
@@ -1,37 +1,83 @@
...
+
+  - stage: test
+    before_script:
+    - echo $DOCKER_TOKEN | docker login --username $DOCKER_USERNAME --password-stdin
+    - docker pull samseedsoftwire/todo-app-test:latest
+    - docker build --target test --tag samseedsoftwire/todo-app-test:latest .
+    script:
+    - docker run samseedsoftwire/todo-app-test:latest tests_unit -rA
+    - docker run samseedsoftwire/todo-app-test:latest tests_integration -rA
+    - docker run -e AUTH_PARAMS_KEY=$AUTH_PARAMS_KEY -e AUTH_PARAMS_TOKEN=$AUTH_PARAMS_TOKEN -e BOARD_ID=$BOARD_ID -e FLASK_APP=$FLASK_APP -e FLASK_ENV=$FLASK_ENV samseedsoftwire/todo-app-test:latest tests_e2e -rA
```
Comment (JackMeadDev) [inline .travis.yml:14]: There's a slightly simpler syntax we can take advantage of here, each env  variable if it's not changing names can be passed through directly like
`... -e AUTH_PARAMS_KEY -e AUTH_PARAMS_TOKEN..` etc.

**Review by JackMeadDev — APPROVED** [review]
This looks good, I can see the app is deploying and working. I've left a few thoughts below, and in particular you should fix up the PORT environment variable so that the prod docker container works according to your readme instructions, but happy for you to merge this once that's working.

With regards to that warning, it's OK to ignore that here. If you did to want to look further into that then [this commit](https://github.com/ghdl/docker/commit/3c7be1d2d8fb6bc4c23b5ab2f3aa0263d6200c30) claims to integrate a docker login step with `pass` but I haven't checked it explicitly

**Pull request closed without merging**